perf: concurrent file import

achingbrain opened the pull request:
Adds two new options:

`fileImportConcurrency`

This controls the number of files that are imported concurrently. You may wish to set this high if you are importing lots of small files.

`blockWriteConcurrency`

This controls how many blocks from each file we hash and write to disk at the same time.  Setting this high when writing large files will significantly increase import speed, though having it high when `fileImportConcurrency` is also high can swamp the process.

'high' is relative to your hardware speed (CPU & disk)..

It also:

1. Flattens module options because validating deep objects was clunky and the separation of access to config sub objects within this module isn't very good
1. Replaces `superstruct` and `deep-extend` with `merge-options` which is better suited for merging options and is smaller
1. Replaces `async-iterator-*` modules with versions from the more zeitgeisty `it-*` namespace

Supersedes #38, sort of.  No batching but atomicity guarantees are maintained and performance gains are broadly similar with the right tuning.

alanshaw commented on package.json:
```diff
@@ -38,33 +38,35 @@
   "dependencies": {
-    "async-iterator-all": "^1.0.0",
-    "async-iterator-batch": "~0.0.1",
-    "async-iterator-first": "^1.0.0",
```
It would be good to get PRs to `js-ipfs` and `js-ipfs-http-client` (and any other projects in our dependency tree) to switch them there also so we don't get multiple of the same thing.

achingbrain replied:
Sure - I'll do that in the PR that adds support for the new options

alanshaw commented on src/dag-builder/file/flat.js:
```diff
@@ -1,11 +1,11 @@
   const roots = []
 
-  for await (const chunk of batch(source, Infinity)) {
+  for await (const chunk of batch(source, Number.MAX_SAFE_INTEGER)) {
```
Just use `it-all`?

alanshaw replied:
...I mean unless you actually want to receive a batch of `Number.MAX_SAFE_INTEGER` items at maximum 🤪

achingbrain replied:
Historically I think the assumption has been that `source` could return variable length arrays so we use a batcher that flattens them first, though the tests seem to pass with a regular batcher so I'm not sure that it's true any more.

In fact I think it only ever worked because of how `Array.concat` treats its arguments (e.g. not rejecting non-array arguments).

```
it-batch (internal buffer uses Array.push and yields when the buffer gets big enough)

input                         batcher                         output
child, child, child, child -> [child, child, child, child] -> [child, child, child, child] 
```

```
it-flat-batch (internal buffer uses Array.concat)

input                             batcher                         output
[child, child, child], [child] -> [child, child, child, child] -> [child, child, child, child] 
```

```
it-all (like it-batch but with an unbounded buffer size)

input                         output
child, child, child, child -> [child, child, child, child]
```

What fun.

TBH I don't know why we have this builder, it's not terribly useful and you can construct the same DAG with the balanced builder and a really high `maxChildrenPerNode` value.

alanshaw commented on README.md:
```diff
@@ -124,10 +124,9 @@ The input's file paths and directory structure will be preserved in the [`dag-pb
-  - `maxChunkSize` (positive integer, defaults to `262144`): the maximum chunk size
+- `avgChunkSize` (positive integer, defaults to `262144`): the average chunk size (rabin chunker only)
+- `minChunkSize` (positive integer): the minimum chunk size (rabin chunker only)
+- `maxChunkSize` (positive integer, defaults to `262144`): the maximum chunk size
```
Please can we get an acccompanying PR to `js-ipfs`?

alanshaw commented on README.md:
```diff
@@ -124,10 +124,9 @@ The input's file paths and directory structure will be preserved in the [`dag-pb
-  - `maxChunkSize` (positive integer, defaults to `262144`): the maximum chunk size
+- `avgChunkSize` (positive integer, defaults to `262144`): the average chunk size (rabin chunker only)
+- `minChunkSize` (positive integer): the minimum chunk size (rabin chunker only)
+- `maxChunkSize` (positive integer, defaults to `262144`): the maximum chunk size
```
`dirBuilder` options (listed later here) have not been flattened...intentional?

achingbrain replied:
Just a mistake.  Internally it's been flattened for ages so the `dirBuilder` options haven't worked for a while.  Oops.

alanshaw commented:
Got any perf stats we can ogle at?

alanshaw commented on README.md:
```diff
@@ -134,9 +134,7 @@ The input's file paths and directory structure will be preserved in the [`dag-pb
-- `dirBuilder` (object): the options for the directory builder
-  - `hamt` (object): the options for the HAMT sharded directory builder
-    - bits (positive integer, defaults to `8`): the number of bits at each bucket of the HAMT
+- `bits` (positive integer, defaults to `8`): the number of bits at each bucket of the HAMT
```
`hamtBucketBits`?

achingbrain commented:
![statto](https://user-images.githubusercontent.com/665810/69722301-ad202200-110e-11ea-8410-55a7db7f8062.jpg)

Data generation, 1000x files 1k-1MB in size filled with random(ish) data:

```console
$ cat gen-data.sh 
#!/bin/bash

mkdir -p data

for n in {1..1000}; do
    dd if=/dev/urandom of=data/file-$( printf %03d "$n" ).bin bs=1 count=$( shuf -i 1024-1048576 -n 1 )
done
```

Generated data:

```console
$ du -hs ./data
525M	./data
```

go-IPFS (IDK why it thinks the data folder is smaller than it is):

```console
$ time ipfs add -r ./data
...
added QmSgH3zFNiMNyH2es2TrkZmjPTETVXvdGQHkA1hPS2n6sL data
 501.99 MiB / 501.99 MiB [============================] 100.00%
real	0m55.654s
user	0m4.931s
sys	0m5.502s
```

js-IPFS Master

```console
$ time jsipfs add -r ./data
added QmSgH3zFNiMNyH2es2TrkZmjPTETVXvdGQHkA1hPS2n6sL data

real	0m52.505s
user	0m13.482s
sys	0m5.760s
```

File & block concurrency = 1 (e.g. how it was before this PR):

```console
$ time jsipfs add -r --file-import-concurrency=1 --block-write-concurrency=1 ./data
...
added QmSgH3zFNiMNyH2es2TrkZmjPTETVXvdGQHkA1hPS2n6sL data

real	0m52.650s
user	0m14.849s
sys	0m5.410s
```

Proposed default values:

```console
$ time jsipfs add -r --file-import-concurrency=50 --block-write-concurrency=10 ./data
...
added QmSgH3zFNiMNyH2es2TrkZmjPTETVXvdGQHkA1hPS2n6sL data

real	0m25.814s
user	0m11.552s
sys	0m5.262s
```
